STM32F3x power management

albayenes opened the pull request:
This patch adds full low-power support to STM32F3 in Zephyr, including:

STOP0 & STOP1 modes (with main/LP regulator)
STANDBY (poweroff) mode
Wakeup pin (PA0) support
SoC clock reinitialization on wakeup
Idle timer using RTC
Samples updated to test wakeup pin and poweroff

This is a significant power-management enablement patch for the STM32F3 series.

dsseng reviewed (commented):
Thank you for your contribution, this is very valuable for other STM32F3 users out there.

Could you please take a look at my comments related to the commit scope and code style?

Also, please sign-off the commits, please find the information in the [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#developer-certification-of-origin-dco)

albayenes commented on samples/boards/st/power_mgmt/blinky/src/main.c:
Fixed

albayenes commented on soc/st/stm32/common/soc_config.c:
```diff
@@ -8,13 +8,13 @@
+#include <stm32_ll_system.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
```
Fixed. I reverted ordering.

dsseng reviewed (commented):
Thanks, looks good and ready for review by the maintainers

etienne-lms commented on soc/st/stm32/stm32f3x/Kconfig.defconfig:
```diff
@@ -14,4 +14,8 @@ configdefault SHARED_INTERRUPTS
 	default y if $(dt_nodelabel_enabled,timers1) && $(dt_nodelabel_enabled,timers16)
 	default y if $(dt_nodelabel_enabled,timers1) && $(dt_nodelabel_enabled,timers17)
 
+config PM
```
Prefer `configdefault`.

etienne-lms replied:
My mistake, forget my comment, it does not apply....

etienne-lms reviewed (commented):
Thanks for the changes. Some comments for consistency among STM32 SoCs.

Nitpicking on the commit message: could you refine the header line, and remove heading indentation in the message body?
```patch
-pm: Add power management for stm32f3x
+soc: st: stm32f3: Add power management
 
-    Adds full low-power support to STM32F3 in Zephyr, including:
+Add full low-power support to STM32F3 in Zephyr, including
 (...)

albayenes commented on soc/st/stm32/stm32f3x/soc.c:
```diff
@@ -31,6 +33,10 @@ void soc_early_init_hook(void)
 	SystemCoreClock = 8000000;
 
+#if CONFIG_PM
+	stm32_power_init();
```
Fixed.

etienne-lms commented on samples/boards/st/power_mgmt/wkup_pins/boards/nucleo_f302r8.overlay:
```diff
@@ -0,0 +1,17 @@
+	gpio_keys {
+		wkup_button: wkup {
+			label = "WKUP";
+			gpios = <&gpioa 0 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
```
Well, I gave it a try. The system powers back when pin A0 (e.g. connector CN7[28]) is connected to 3V3 supply (e.g. CN7[16]). From reference manual, WKUP pins wake the system on rising edge only.
```suggestion
			/*
			 * PA0 is not connected to any button: you must connect it to Vdd (+3.3V)
			 * using a jumper wire. For example, wiring PA0 CN7[28] (PA0) to CN7[16]
			 */
			gpios = <&gpioa 0 (GPIO_ACTIVE_HIGH | GPIO_PULL_DOWN)>;

albayenes replied:
Yes there is no push button connected to GPIOA0 on nucleo board F3x. I tested on F302R8. I added your suggestion.

etienne-lms commented on soc/st/stm32/stm32f3x/soc.c:
```diff
@@ -31,6 +33,10 @@ void soc_early_init_hook(void)
 	/* At reset, system core clock is set to 8 MHz from HSI */
 	SystemCoreClock = 8000000;
 
+#if defined(CONFIG_PM) || defined(CONFIG_POWEROFF)
```
Sorry for my previous comment.
```suggestion
#if defined(CONFIG_PM)
```
unless what build fails when PM is disabled and POWEROFF is enabled.

etienne-lms commented on samples/boards/st/power_mgmt/wkup_pins/boards/nucleo_f302r8.overlay:
```diff
@@ -0,0 +1,21 @@
+/ {
```
Could you add a copyright/license notice inline comment as requested by CI compliance checks?

albayenes replied:
Fixed.

etienne-lms commented on soc/st/stm32/stm32f3x/Kconfig.defconfig:
```diff
@@ -14,4 +14,14 @@ configdefault SHARED_INTERRUPTS
+if PM
+
+config COUNTER
+        default y
```
Could you address CI compliance checks requests (a tab instead of space chars)?

I'm surprised you experience build issues using `configudefault` instead of `config` in this file.

albayenes commented on dts/arm/st/f3/stm32f3.dtsi:
```diff
@@ -31,8 +33,27 @@
+				compatible = "zephyr,power-state";
+				power-state-name = "suspend-to-idle";
+				substate-id = <1>;
+				min-residency-us = <1000000>;
```
Fixed.

albayenes commented on samples/boards/st/power_mgmt/wkup_pins/boards/nucleo_f302r8.overlay:
```diff
@@ -0,0 +1,21 @@
+			label = "WKUP";
+			/*
+			 * PA0 is not connected to any button: you must connect it to Vdd (+3.3V)
+			 * using a jumper wire. For example, wiring PA0 CN7[28] (PA0) to CN7[16]
```
Fixed.

etienne-lms reviewed (commented):
2 nitpicking comments. Otherwise LGTM.

etienne-lms reviewed (commented):
LGTM with a last comment preferably to address.

albayenes commented on soc/st/stm32/stm32f3x/Kconfig.defconfig:
```diff
@@ -14,4 +14,14 @@ configdefault SHARED_INTERRUPTS
 
+if PM
+
+config COUNTER
```
You are right, it is compiled successfully.

mathieuchopstm reviewed (dismissed):
NAK for `k_cpu_idle` usage, seems fair otherwise.

Nit: please split implementation of `poweroff` and `power` (low-power) in two separate commits.

erwango commented:
@albayenes Do you plan to continue this PR ?

dsseng reviewed (commented):
Mostly good, please fix the compliance errors CI reports

etienne-lms reviewed (dismissed):
LGTM

albayenes committed:
soc: st: stm32f3: Add low-power mode support

Adds full low-power support to STM32F3 in Zephyr, including:

STOP0 & STOP1 modes (with main/LP regulator)
Wakeup pin (PA0) support
SoC clock reinitialization on wakeup
Idle timer using RTC
Samples updated to test wakeup pin

This is a significant power-management enablement
patch for the STM32F3 series.

Tested on: NUCLEO-F302R8 board.

Signed-off-by: Enes Albay <albayenes@gmail.com>

albayenes committed:
soc: st: stm32f3: Add poweroff support

Tested on: NUCLEO-F302R8 board.

Signed-off-by: Enes Albay <albayenes@gmail.com>

albayenes commented on soc/st/stm32/stm32f3x/poweroff.c:
```diff
@@ -0,0 +1,20 @@
+	LL_PWR_ClearFlag_SB();
+	LL_PWR_ClearFlag_WU();
+
+	LL_PWR_SetPowerMode(LL_PWR_MODE_STANDBY);
```
Fixed.

albayenes commented:
> @albayenes, could you rebase your 2 commits on mainline `main` branch instead of appending a merge commit in your patch series?

Fixed.

albayenes commented on soc/st/stm32/stm32f3x/power.c:
```diff
@@ -0,0 +1,72 @@
+		break;
+
+	default:
+		LOG_WRN("Unsupported power substate-id %u", state);
```
Fixed.

albayenes commented on soc/st/stm32/stm32f3x/power.c:
```diff
@@ -0,0 +1,72 @@
+	/* Enable Power clock. It should by done by default, but make sure to
+	 * enable it for all STM32F3x chips.
+	 */
+	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
```
Fixed

albayenes commented:
Could you reopen the PR?

mathieuchopstm commented:
> Could you reopen the PR?

It doesn't appear to be possible:
<img width="382" height="132" alt="image" src="https://github.com/user-attachments/assets/719248a4-69c1-4a4a-ab68-7312674af6de" />

Please create a new PR.

erwango commented:
> Could you reopen the PR?

@albayenes It's probably because branch albayenes:stm32f3_pm was updated while PR was closed. It **may** be possible to open it again if you align branch on PR state again.
